File: design-pattern/template-pattern/template.cpp
#include <iostream>
#include <string>

using namespace std;

class modelTrainer
{
public:
    void executeTrainingPipeline(string &data)
    {
        collectData(data);
        preprocessData();
        trainModel();
        evaluateModel();
        saveModel();
    }

protected:
    void collectData(string &data)
    {
        cout << "Collecting data for Model Training" << endl;
    };

    void preprocessData()
    {
        cout << "Preprocessing data for Model Training" << endl;
    };

    virtual void trainModel() = 0;
    virtual void evaluateModel() = 0;

    void saveModel()
    {
        cout << "Saving Model after Training" << endl;
    };
};

class LinearRegressionTrainer : public modelTrainer
{
public:
    void trainModel() override
    {
        cout << "Training Linear Regression Model with data: " << endl;
    };

    void evaluateModel() override
    {
        cout << "Evaluating Linear Regression Model" << endl;
    };
};

class NLPTrainer : public modelTrainer
{
public:
    void trainModel() override
    {
        cout << "Training NLP Model with data: " << endl;
    };

    void evaluateModel() override
    {
        cout << "Evaluating NLP Model" << endl;
    };
};

class SVMTrainer : public modelTrainer
{
public:
    void trainModel() override
    {
        cout << "Training SVM Model with data: " << endl;
    };

    void evaluateModel() override
    {
        cout << "Evaluating SVM Model" << endl;
    };
};

int main(void)
{
    string data = "Sample Training Data";

    modelTrainer *trainer = new NLPTrainer();
    trainer->executeTrainingPipeline(data);
    delete trainer;

    trainer = new SVMTrainer();
    trainer->executeTrainingPipeline(data);
    delete trainer;
};

File: design-pattern/template-pattern/template.ts

abstract class modelTrainer {

    templateMethod(dataPath: string, modelPath: string): void {
        this.loadData(dataPath);
        this.preprocessData();
        this.trainModel();
        this.evaluateModel();
        this.saveModel(modelPath);
    }

    protected loadData(dataPath: string): void {
        console.log("Loading data...");
    }

    protected preprocessData(): void {
        console.log("Preprocessing data...");
    }

    protected abstract trainModel(): void;
    protected abstract evaluateModel(): void;

    protected saveModel(modelPath: string): void {
        console.log("Saving model...");
    }
}

class decisionTreeTrainer extends modelTrainer {
    trainModel(): void {
        console.log("Training Decision Tree model...");
    }

    evaluateModel(): void {
        console.log("Evaluating Decision Tree model...");
    }
};

class neuralNetworkTrainer extends modelTrainer {
    trainModel(): void {
        console.log("Training Neural Network model...");
    };

    evaluateModel(): void {
        console.log("Evaluating Neural Network model...");
    }
};

const algo = new decisionTreeTrainer();
algo.templateMethod("data/train.csv", "models/decision_tree.model");